### PR TITLE
TLSv1.3: Interoperable 1-RTT handshake

### PR DESCRIPTION
**Pull request by mattcaswell**
##### Checklist

- [x] tests are added or updated
- [x] CLA is signed

##### Description of change

The objective of this PR is to get a simple, inter-operable TLSv1.3 1-RTT handshake working. It is currently marked as WIP, primarily because I need to add a number of tests to this.

Some caveats:
- By interoperable I mean we can complete a simple 1-RTT handshake with picotls, with OpenSSL either as the client or the server. I have not tried it with other libraries.
- This does not support HelloRetryRequest yet. If we encounter one of those the connection will fail.
- No resumption. NewSessionTicket messages are ignored.
- We currently only send one key_share. By default this is for X25519 (although this can be changed). Picotls currently expects a P-256 key share by default so we need to work around that for now.
- No client auth. Connections will fail if you try it.
- This only supports RSA certificates at the moment

To try it out clone picotls and build it according to the instructions:
https://github.com/h2o/picotls

To run picotls as a server:

````
./cli -c <certfile> -k <keyfile> 127.0.0.1 8443
````
and OpenSSL as the client:
````
openssl s_client -connect 127.0.0.1:8443 -curves P-256
````

To run OpenSSL as the server:
````
openssl s_server -named_curve P-256 -accept 127.0.0.1:8443
````
And picotls as the client:
````
./cli 127.0.0.1 8443
````

### CODE REVIEW

**Comment by davidben** [comment]
Testing against BoringSSL deployments, I noticed two problems:
* You send {3, 4} in the record layer rather than {3, 1}. TLS 1.3 freezes the record-layer version at {3, 1}. (If I fix this, it interops with BoringSSL though! :-D)
* This PR breaks TLS 1.2. The WG decided that, by advertising RSA-PSS, you are also offering to do it in TLS 1.2. The new signature algorithm setup is meant to be backported. Google servers already deploy RSA-PSS at TLS 1.2. Chrome and Firefox's upcoming releases will as well.

The BoringSSL test suite would have caught all of this, but I haven't tried it yet. (Update to the latest revision first. The one you're at implement draft 16, not draft 18.)

We've also got enough of a deployment by now you don't need to do anything complicated to test against it:
* For a server, connect to https://cert-test.sandbox.google.com. (Note this server mandates SNI.)
* For a client, use Chrome Canary (or Dev or Beta) and it'll be on with 50% probability. Go to chrome://flags/#ssl-version-max to force it on.

For the second, perhaps an intermediate step to make the signature algorithms stuff properly negotiated? Teach TLS 1.2 and proto-TLS-1.3 about RSA-PSS, then teach proto-TLS-1.3 that PKCS#1 is forbidden and about ECDSA's hash/curve matching. We actually did that before changing the handshake since it was a nice self-contained change like the record layer.

**Comment by mattcaswell** [comment]
@davidben Awesome! Thanks for the feedback...I'll work on fixing those issues.

**Comment by mattcaswell** [comment]
Branch updated to add various tests. Also fixes the issues identified by @davidben. I have successfully managed to create TLSv1.3 connections with the google sandbox server as well as Chrome Canary. Once #2020 goes in, I will remove the WIP label from this PR.

**Comment by mattcaswell** [comment]
Now that #2020 has gone in, I have rebased this and removed the WIP label. This is now ready for review.

**Comment by richsalz** [comment]
Please look at #2052

**Comment by martinthomson** [comment]
FWIW, this worked with NSS first time.  Tested:
* `openssl s_client -connect 127.0.0.1:4433` against `selfserv -d ~/ssl_gtests -n rsa2048 -p 4433 -V tls1.3: -v`
* `tstclnt -h 127.0.0.1 -p 4433 -V tls1.3: -v -o -d ~/ssl_gtests` against `openssl s_server -accept 127.0.0.1:4433`
Where `server.pem` contains a cert and key for openssl and `~/ssl_gtests` is an NSS database containing an RSA key and certificate called 'rsa2048'.

I tried P-256, x25519 and P-384, which all worked.  ECDSA (openssl s_client) fails as expected.  Not a lot else to try based on the description.  If there is anything you want me to poke at, let me know and I can give it a spin.

Code inspection suggests that there are plenty of work left to do, but this is looking pretty promising.

**Comment by mattcaswell** [comment]
I rebased this. Also added a couple of commits with tweaks for Travis/Appveyor issues. Let's see how travis/appveyor do.

File: ssl/t1_lib.c [hunk 1]
```diff
@@ -719,55 +719,94 @@ static int tls1_check_cert_param(SSL *s, X509 *x, int set_ee_md)
...
+    unsigned int sigalg;
+    int hash;
+    int sig;
+} SIGALG_LOOKUP;
+
+static SIGALG_LOOKUP sigalg_lookup_tbl[] = {
+    {TLSEXT_SIGALG_ecdsa_secp256r1_sha256, NID_sha256, EVP_PKEY_EC},
+    {TLSEXT_SIGALG_ecdsa_secp384r1_sha384, NID_sha384, EVP_PKEY_EC},
+    {TLSEXT_SIGALG_ecdsa_secp521r1_sha512, NID_sha512, EVP_PKEY_EC},
+    {TLSEXT_SIGALG_ecdsa_sha1, NID_sha1, EVP_PKEY_EC},
```
Comment (levitte) [inline ssl/t1_lib.c]: You should probably do some `#ifndef OPENSSL_NO_EC` wrapping...

File: ssl/t1_lib.c [hunk 2]
```diff
@@ -719,55 +719,94 @@ static int tls1_check_cert_param(SSL *s, X509 *x, int set_ee_md)
...
+    {TLSEXT_SIGALG_rsa_pss_sha384, NID_sha384, EVP_PKEY_RSA},
+    {TLSEXT_SIGALG_rsa_pss_sha512, NID_sha512, EVP_PKEY_RSA},
+    {TLSEXT_SIGALG_rsa_pkcs1_sha256, NID_sha256, EVP_PKEY_RSA},
+    {TLSEXT_SIGALG_rsa_pkcs1_sha384, NID_sha384, EVP_PKEY_RSA},
+    {TLSEXT_SIGALG_rsa_pkcs1_sha512, NID_sha512, EVP_PKEY_RSA},
+    {TLSEXT_SIGALG_rsa_pkcs1_sha1, NID_sha1, EVP_PKEY_RSA},
+    {TLSEXT_SIGALG_dsa_sha256, NID_sha256, EVP_PKEY_DSA},
+    {TLSEXT_SIGALG_dsa_sha384, NID_sha384, EVP_PKEY_DSA},
+    {TLSEXT_SIGALG_dsa_sha512, NID_sha512, EVP_PKEY_DSA},
+    {TLSEXT_SIGALG_dsa_sha1, NID_sha1, EVP_PKEY_DSA},
```
Comment (levitte) [inline ssl/t1_lib.c]: `#ifndef OPENSSL_NO_DSA` wrapping?

**Comment by mattcaswell** [comment]
Grrr...more travis failure fixes pushed.
@levitte: also fixed no-ec/no-dsa. For some reason I can't see your comment in this PR, although I saw it in my email.

**Comment by levitte** [comment]
> @levitte: also fixed no-ec/no-dsa. For some reason I can't see your comment in this PR, although I saw it in my email.

I thought I was wrong in those comments, so I removed them.  Thanks for still fixing

**Comment by mattcaswell** [comment]
> I thought I was wrong in those comments, so I removed them. Thanks for still fixing

You weren't!

**Comment by richsalz** [comment]
+1 reconfirmed

**Comment by mattcaswell** [comment]
And.....pushed to master!!! Thanks @richsalz, @levitte, @davidben and @martinthomson for reviewing and testing this PR!

**Pull request closed without merging**